src/accelerate/utils/numactl.py

```diff
@@ -0,0 +1,67 @@
+# Copyright 2024 The HuggingFace Team. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import os
+import subprocess
+import sys
+
+from accelerate.utils.environment import _get_nvidia_smi
+
+
+def get_query_command(device_id: int = None, backend: str = None):
+    """
+    Returns the right param args for `numa` support based on a `backend`.
+
+    Currently only supports `nvidia`.
+    """
+    if backend == "nvidia":
+        return [_get_nvidia_smi(), "--query-gpu=pci.bus_id", "-i", str(device_id), "--format=csv,noheader"]
+    else:
+        # TODO: Add support for AMD and Gaudi2
+        raise ValueError(f"Backend {backend} is currently not supported through `accelerate launch`.")
+
+
+def get_bus_id(device_id: int = None, backend: str = None):
+    """
+    Gets the bus ID for `device_id`. If not passed, will default to the local rank.
+    """
+    if device_id is None:
+        device_id = os.environ.get("LOCAL_RANK", 0)
+    command = get_query_command(device_id, backend)
+    output = subprocess.check_output(command, universal_newlines=True)
+    return output.strip()[4:]
+
+
+def get_numa_node_for_device(device_id: int = None):
+    """
+    Gets the numa node for `device_id`. If not passed, will default to the local rank.
+    """
+    bus_id = get_bus_id(device_id)
+    if bus_id is None:
+        bus_id = 4
```

Comment:
Could you please add a comment where `4` is coming from?

Reply:
cc @stas00 for insight? 👀

Reply:
yeah, the above is not what the bash code did. Let's look at the original bash:

```
# Query the bus ID for device LOCAL_RANK
BUS_ID=$(nvidia-smi --query-gpu=pci.bus_id -i $LOCAL_RANK --format=csv,noheader)
BUS_ID=${BUS_ID,,}

# Find the numa node for device LOCAL_RANK
NODE=$(cat /sys/bus/pci/devices/${BUS_ID:4}/numa_node)
```
4 is there to remove the first 4 characters - it's indeed a good idea to add a comment explaining it, perhaps copy this verbatim as it's self-explanatory:

```
$ LOCAL_RANK=5 BUS_ID=$(nvidia-smi --query-gpu=pci.bus_id -i $LOCAL_RANK --format=csv,noheader)
$ echo $BUS_ID
00000000:85:00.0
$ ls -l /sys/bus/pci/devices/${BUS_ID}
ls: cannot access '/sys/bus/pci/devices/00000000:85:00.0': No such file or directory
$ echo ${BUS_ID:4}
0000:85:00.0
$ ls -l /sys/bus/pci/devices/${BUS_ID:4}
lrwxrwxrwx 1 root root 0 Feb 28 19:38 /sys/bus/pci/devices/0000:85:00.0 -> ../../../devices/pci0000:81/0000:81:00.0/0000:82:00.0/0000:83:01.0/0000:85:00.0
```

`${var:pos}` is bash's adoption of ksh functionality - it'll truncate the string starting at `pos`